### PR TITLE
Fix CMake's install prefix to always use one location

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
Otherwise the python bindings break.

The python file: `bindings/py/packaging/setup.py` line: 38 will break.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -45,7 +45,7 @@ if(NOT CMAKE_BUILD_TYPE)
 endif()
 
 if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
-   set(CMAKE_INSTALL_PREFIX "${PROJECT_SOURCE_DIR}/build/${CMAKE_BUILD_TYPE}" CACHE PATH "Default Install location" FORCE)
+   set(CMAKE_INSTALL_PREFIX "${PROJECT_SOURCE_DIR}/build/Release" CACHE PATH "Default Install location" FORCE)
```
Comment (breznak) [inline CMakeLists.txt]: What breaks, is the variable not passed correctly? 

I think @dkeeney used the "variable build type" for MS IDE (which apparently runs in parallel Release & Debug builds?)

Comment (ctrl-z-9000-times) [reply CMakeLists.txt]: This line breaks:
https://github.com/htm-community/nupic.cpp/blob/0048e5cd90053ce2f4c205bc077fee15e5e46b3d/bindings/py/packaging/setup.py#L38

Comment (dkeeney) [reply CMakeLists.txt]: Yes, for MSVC you need different folders for Release and Debug.
My thought was that setup.py would build only the Release version.  If you need to build a Debug version for the python interface then you have to make a few changes manually.  (change the build type) Is there a need for that?

Comment (breznak) [reply CMakeLists.txt]: >  DISTR_DIR = os.path.join(REPO_DIR, "build", "Release", "distr") 

Sorry, I looked at that line but then got confused by the git diff. 
So, can we fix the setup.py instead? 
 - expose ENV var from Cmake, read it in setup? 
 - try both Debug/Release folders, whichever exists? (problem, if both do and one is old) 
 - setup py always builds in Release, so pass to that to underliyng cmake? 

- one of our CI runs in Debug, how come it's not broken?

Comment (ctrl-z-9000-times) [reply CMakeLists.txt]: > Is there a need for that?

It would be useful for debugging the python interface, as well as all of the C++ code which is called by the python interface.

Comment (breznak) [reply CMakeLists.txt]: > My thought was that setup.py would build only the Release version.  

I'm fine with this :+1: 
- The behavior should be documented in Readme ("Calling python setup.py always results in Release build." 
- is our Debug CI correct?
OSX CI runs in Debug: circle.yml :  
```
cd build/scripts
cmake ../.. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../Release
make | grep -v -F '\\-\\- Installing:'
make install 2>&1 | grep -v -F 'Installing:'
- run:
name: Install Python extension
command: python setup.py install --user --prefix=
```
does it use the "hack" and actually run in Debug (seems so from the catched bugs), or does it rebuild in Release? 

TL;DR: likely not an issue, but intended behavior. 
> cmake ../.. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../Release 

use the following hack to run python setup with Debug. 

TODO: 
- [ ] please add a doc to Readme (setup -> Release) 
- [ ] add a comment to the line 38 
` DISTR_DIR = os.path.join(REPO_DIR, "build", "Release", "distr") `
- [ ] and to the OSX CI 
>  cmake ../.. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../Release 

So we don't "correct" it by mistake

Comment (dkeeney) [reply CMakeLists.txt]: >  or does it rebuild in Release?

Yes, it will rebuild in Release.

> use the following hack to run python setup with Debug.
cmake ../.. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../Release

Yes this is the hack needed (but include -DBINDING_BUILD=Python3 )
 but it needs to be called  in line 211 of bindings/py/packaging/setup.py.  There are two calls there, the first will do the above hack and set the build type in the cache and the second one essentially does a `make install`.

It would be nice to change this so that it can be driven by an option if you would like to take on that challenge.

There are two setup.py files.   The one at the top of the repository does little except to call the other.  The one in packaging needs to be there so that it ends up in the distribution.

**Comment by breznak** [comment]
I'm closing the PR, if someone wants to work on the new direction, please reopen.
